stdlib/public/SDK/Foundation/TypePreservingNSNumber.mm

```diff
@@ -0,0 +1,229 @@
+    return;
+  case SwiftUInt:
+    memcpy(value, self->storage, sizeof(NSUInteger));
+    return;
+  case SwiftFloat:
+    memcpy(value, self->storage, sizeof(float));
+    return;
+  case SwiftDouble:
+    memcpy(value, self->storage, sizeof(double));
+    return;
+  case SwiftCGFloat:
+    memcpy(value, self->storage, sizeof(CGFloat));
+    return;
+  case SwiftBool:
+    memcpy(value, self->storage, sizeof(bool));
+    return;
+  }
+  swift::swift_reportError(
+      /* flags = */ 0,
+      "_SwiftTypePreservingNSNumber.tag is corrupted.\n");
+}
+
+#define DEFINE_ACCESSOR(C_TYPE, METHOD_NAME) \
+  - (C_TYPE)METHOD_NAME { \
+    switch(tag) { \
+    case SwiftInt: { \
+      NSInteger result; \
+      memcpy(&result, self->storage, sizeof(result)); \
+      return result; \
+    } \
+    case SwiftUInt: { \
+      NSUInteger result; \
+      memcpy(&result, self->storage, sizeof(result)); \
+      return result; \
+    } \
+    case SwiftFloat: { \
+      float result; \
+      memcpy(&result, self->storage, sizeof(result)); \
+      return result; \
+    } \
+    case SwiftDouble: { \
+      double result; \
+      memcpy(&result, self->storage, sizeof(result)); \
+      return result; \
+    } \
+    case SwiftCGFloat: { \
+      CGFloat result; \
+      memcpy(&result, self->storage, sizeof(result)); \
+      return result; \
+    } \
+    case SwiftBool: { \
+      bool result; \
+      memcpy(&result, self->storage, sizeof(result)); \
+      return result; \
+    } \
+    } \
+    swift::swift_reportError( \
+        /* flags = */ 0, \
+        "_SwiftTypePreservingNSNumber.tag is corrupted.\n"); \
+  }
+
+DEFINE_ACCESSOR(char, charValue)
+DEFINE_ACCESSOR(int, intValue)
+DEFINE_ACCESSOR(unsigned int, unsignedIntValue)
+DEFINE_ACCESSOR(long long, longLongValue)
+DEFINE_ACCESSOR(unsigned long long, unsignedLongLongValue)
+DEFINE_ACCESSOR(float, floatValue)
+DEFINE_ACCESSOR(double, doubleValue)
+
+#undef DEFINE_ACCESSOR
+
+- (Class)classForCoder {
+  return [NSNumber class];
+}
+
+#define DEFINE_INIT(C_TYPE, FUNCTION_NAME) \
+  extern "C" NS_RETURNS_RETAINED NSNumber * \
+  _swift_Foundation_TypePreservingNSNumberWith ## FUNCTION_NAME(C_TYPE value) { \
+    _SwiftTypePreservingNSNumber *result = \
+      [[_SwiftTypePreservingNSNumber alloc] init]; \
+    result->tag = Swift ## FUNCTION_NAME; \
+    memcpy(result->storage, &value, sizeof(value)); \
+    return result; \
+  }
+
+DEFINE_INIT(NSInteger, Int)
+DEFINE_INIT(NSUInteger, UInt)
+DEFINE_INIT(float, Float)
+DEFINE_INIT(double, Double)
+DEFINE_INIT(CGFloat, CGFloat)
+
+extern "C" NS_RETURNS_RETAINED NSNumber *
+_swift_Foundation_TypePreservingNSNumberWithBool(uint8_t value) {
+  _SwiftTypePreservingNSNumber *result =
+    [[_SwiftTypePreservingNSNumber alloc] init];
+  result->tag = SwiftBool;
+  bool boolValue = value;
+  memcpy(result->storage, &boolValue, sizeof(boolValue));
+  return result;
+}
+
+#undef DEFINE_INIT
+
+extern "C" uint32_t
```

Comment:
Please mark these as SWIFT_ABI, or whatever we're using for `silgen_name` functions.